### DISCUSSION TITLE
Assessment of the difficulty in porting CPU architecture for SuiteSparse

### DISCUSSION THREAD

**Discussion by wangyuliu** [DrTimothyAldenDavis/SuiteSparse#525] [General]
Hello everyone! I am working on implementing a tool to assess the complexity of CPU architecture porting. It primarily focuses on RISC-V architecture porting. In fact, the tool may have an average estimate of various architecture porting efforts.My focus is on the overall workload and difficulty of transplantation in the past and future,even if a project has already been ported.As part of my dataset, I have collected the **SuiteSparse** project. **I would like to gather community opinions to support my assessment. I appreciate your help and response!** Based on scanning tools, the porting complexity is determined to be moderate leaning towards simple, with a moderate amount of code related to the CPU architecture in the project.  Is this assessment accurate?Do you have any opinions on personnel allocation and consumption time？ I look forward to your help and response.

**Comment by DrTimothyAldenDavis** [comment]
I would guess this is the case.  We already have workflows that compile and test SuiteSparse on many architectures and OS's  (Linux, Mac, Windows with MSVC, and Windows with MING).  The arch-emu workflow runs on x86, aarch64, armv7, ppc64le, and s390x.

So it's likely a RISC-V port would be simple.  It would have to have BLAS, LAPACK (for UMFPACK, CHOLMOD, and SPQR), and for SPEX it would need GMP and MPRF.

The copy I have of METIS 5.1 in CHOLMOD/SuiteSparse_metis can sometimes give portability issues, but it's a subset of METIS that I use.  I don't build the METIS programs, for example, which uses regex.h, which causes Windows portability problems.

**Reply by mmuetzel** [reply]
When it comes to CI for RISC-V, see this upstream tracker issue:
https://gitlab.alpinelinux.org/alpine/aports/-/issues/13269

IIUC, they are waiting for hardware actually capable of handling their build process. Once they'd release packages for that platform, it would probably be straight-forward to add a CI runner (using the existing action that uses `qemu`).

**Comment by wangyuliu** [comment]
Thank you very much for your answer! Our tool has obtained a portion of the assembly and Intrinsic code from suiesparse, which is not very extensive. We believe that this requires some effort when porting to the CPU architecture, so evaluating the difficulty of porting is moderate to simple. Is this correct?is this evaluation strategy effective?Looking forward to your further response.

**Reply by DrTimothyAldenDavis** [reply]
I don't have any additional insight into the question.
